test/test_nn.py

```diff
@@ -1938,6 +1939,195 @@ def test_vector_to_parameters(self):
+                return torch.cat([x.new_zeros(1), x[1:]])
+
+        class LastZero(P.Parametrization):
+            def forward(self, x):
+                return torch.cat([x[:-1], x.new_zeros(1)])
+
+        model = nn.Linear(8, 8)
+        initial_weight = model.weight
+        initial_bias = model.bias
+        initial_model = deepcopy(model)
+
+        # Test one parametrization
+        P.register_parametrization(model, "weight", Skew())
+        self.assertTrue(hasattr(model, "parametrizations"))
+        self.assertTrue(P.is_parametrized(model))
+        self.assertTrue(P.is_parametrized(model, "weight"))
+        self.assertNotIn("weight", model._parameters)
+        # Result should be skew-symmetric
+        A = model.weight
+        self.assertAlmostEqual(torch.norm(A + A.t()), 0.)
+        # Remove and check consistency
+        P.remove_parametrization(model, "weight", leave_parametrized=False)
+        self.assertFalse(hasattr(model, "parametrizations"))
+        self.assertEqual(model.weight, initial_model.weight)
+        self.assertEqual(id(model.weight), id(initial_weight))
+
+        # Test two parametrizations at the same time and removing them
+        P.register_parametrization(model, "weight", Skew())
+        P.register_parametrization(model, "weight", Orthogonal())
+        # Result should be orthogonal
+        X = model.weight
+        Id = torch.eye(X.size(0), device=X.device)
+        self.assertAlmostEqual(torch.norm(torch.mm(X.t(), X) - Id), 0.)
+        # Result should be skew-symmetric
+        A = model.parametrizations.weight.original
+        self.assertAlmostEqual(torch.norm(A + A.t()), 0.)
+        # Structure tests
+        self.assertTrue(hasattr(model, "parametrizations"))
+        self.assertTrue(P.is_parametrized(model))
+        self.assertTrue(P.is_parametrized(model, "weight"))
+        self.assertFalse(P.is_parametrized(model, "bias"))
+        self.assertIn("weight", model.parametrizations)
+        self.assertNotIn("weight", model._parameters)
+        p1 = model.parametrizations.weight
+        self.assertTrue(hasattr(p1, "parametrizations"))
+        self.assertTrue(P.is_parametrized(p1))
+        self.assertTrue(P.is_parametrized(p1, "original"))
+        self.assertNotIn("original", p1._parameters)
+        p2 = p1.parametrizations.original
+        self.assertFalse(hasattr(p2, "parametrizations"))
+        self.assertFalse(P.is_parametrized(p2))
+        self.assertIn("original", p2._parameters)
+        # Remove
+        P.remove_parametrization(model, "weight", leave_parametrized=False)
+        self.assertEqual(model.weight, initial_model.weight)
+        self.assertEqual(id(model.weight), id(initial_weight))
+        self.assertFalse(hasattr(model, "weight_cache"))
+        self.assertFalse(hasattr(model, "parametrizations"))
+
+        # Add everything
+        P.register_parametrization(model, "weight", Skew())
+        P.register_parametrization(model, "weight", Orthogonal())
```

Comment:
A question: can we do this reparameterization explicit e.g:
```python
model.weight = Skew(Orthogonal(model.weight))
```
?

Reply:
At the moment this is not possible, as we need to modify the Module `model` to be able to register the parametrizations, although I agree that it would be awesome to have an API like this.

Reply:
I guess that it could be done modifying the `__setattr__` in `nn.Module`, but at least for the first iteration @apaszke said that modifying `nn.Module` was off the table.

Reply:
a middle ground could be `model.register_parametrization("weight", nn.Sequential(Orthogonal(), Skew()))`. I think explicit chaining is better than implicit chaining - and more amenable to programmatic modifications / pretty printing etc

Reply:
We are discussing exactly that point Adam and me. It is true that explicit is better than implicit, but it is not very clear how to make that work with having the parametrizations having control over the parameter itself.

In particular, we would want parametrizations to be able to evaluate the current chain of parametrizations up to them (this is useful to implement parametrizations that change according to certain conditions) and be able to modify the parameter (for initialization purposes).

The `Sequential` approach does not give the items that it contains enough information to implement these things.

Reply:
Maybe another special container for parametrizations can be implemented `nn.ParametrizationList(...)`?

The initialization concern cannot be solved in the general case :(, since manual initialization is tied to knowing the parametrization used.

If torch.Tensor / nn.Parameter were supported evaluation through some virtual method (e.g. `forward()` / `__call__`), then reparametrization buffers and reinitialization could be attached to existing Tensor / Parameter, but this would require modifying basic API / structures (analagous to modifying Module's `__setattr__` / `__getattr__`)

Also one useful design question: will it support well functional interface? (where no modules are used at all)

Reply:
I agree with the attachment to `Tensor` and `Parameter`, as it was discussed in #7313, but it would be quite a big change. Maybe it is more reasonable to wait and see how this way to perform constrained optimisation is adopted, and if people like it, we could revisit it.

At the moment, parametrizations depend on `Modules` in a fundamental way, as they basically inject a `property` on the module in place of the weight, so I don't think that they would work well with the functional interface...

Reply:
> At the moment, parametrizations depend on `Modules` in a fundamental way, as they basically inject a `property` on the module in place of the weight, so I don't think that they would work well with the functional interface...

sigh :(